tooling/src/hypothesistooling/projects/conjecturerust.py

```diff
@@ -0,0 +1,126 @@
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import os
+import subprocess
+
+import hypothesistooling as tools
+import hypothesistooling.installers as install
+import hypothesistooling.releasemanagement as rm
+from hypothesistooling import git
+from hypothesistooling.junkdrawer import in_dir
+
+PACKAGE_NAME = 'conjecture-rust'
+
+CONJECTURE_RUST = os.path.join(tools.ROOT, PACKAGE_NAME)
+
+BASE_DIR = CONJECTURE_RUST
+
+TAG_PREFIX = PACKAGE_NAME + '-'
+
+RELEASE_FILE = os.path.join(BASE_DIR, 'RELEASE.md')
+CHANGELOG_FILE = os.path.join(BASE_DIR, 'CHANGELOG.md')
+GEMSPEC_FILE = os.path.join(BASE_DIR, 'hypothesis-specs.gemspec')
+
+CARGO_FILE = os.path.join(BASE_DIR, 'Cargo.toml')
+
+SRC = os.path.join(BASE_DIR, 'lib')
+
+
+def has_release():
+    """Is there a version of this package ready to release?"""
+    return os.path.exists(RELEASE_FILE)
+
+
+def update_changelog_and_version():
+    """Update the changelog and version based on the current release file."""
+    release_type, release_contents = rm.parse_release_file(RELEASE_FILE)
+    version = current_version()
+    version_info = rm.parse_version(version)
+
+    version, version_info = rm.bump_version_info(version_info, release_type)
+
+    rm.replace_assignment(CARGO_FILE, 'version', repr(version))
+
+    rm.update_markdown_changelog(
+        CHANGELOG_FILE,
+        name='Conjecture for Rust',
+        version=version,
+        entry=release_contents,
+    )
+    os.unlink(RELEASE_FILE)
+
+
+def commit_pending_release():
+    """Create a commit with the new release."""
+    git('rm', RELEASE_FILE)
+    git('add', CHANGELOG_FILE, GEMSPEC_FILE)
+
+    git(
+        'commit', '-m',
+        'Bump conjecture-rust version to %s and update changelog'
+        '\n\n[skip ci]' % (current_version(),)
+    )
+
+
+def cargo(*args):
+    install.ensure_rustup()
+    with in_dir(BASE_DIR):
+        subprocess.check_call(('cargo',) + args)
+
+
+def build_distribution():
+    """Build the rubygem."""
+    cargo('package')
+
+
+def tag_name():
+    """The tag name for the upcoming release."""
+    return TAG_PREFIX + current_version()
+
+
+def has_source_changes():
+    """Returns True if any source files have changed."""
+    return tools.has_changes([SRC])
+
+
+def current_version():
+    """Returns the current version as specified by the gemspec."""
+    return rm.extract_assignment(CARGO_FILE, 'version')
+
+
+CARGO_CREDENTIALS = os.path.expanduser('~/.cargo/credentials')
+
+
+def upload_distribution():
+    """Upload the built package to rubygems."""
+    tools.assert_can_release()
+
+    # Yes, cargo really will only look in this file. Yes this is terrible.
+    # This only runs on Travis, so we may be assumed to own it, but still.
+    os.path.unlink(CARGO_CREDENTIALS)
```

Comment:
This will fail if credentials aren’t already set – e.g. if the cache has just been purged. Another addition to the junk drawer!

```python
import errno
import os


def rm_f(path):
    try:
        os.path.unlink(path)
    except OSError as err:
        if err.errno == errno.ENOENT:
            pass
        else:
            raise
```